tracing/tracing.go

```diff
@@ -0,0 +1,136 @@
+type fileExporter struct {
+	file           *os.File
+	writerExporter *stdouttrace.Exporter
+}
+
+var _ trace.SpanExporter = &fileExporter{}
+
+func newFileExporter(file string) (*fileExporter, error) {
+	f, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		return nil, fmt.Errorf("opening %s: %w", file, err)
+	}
+	stdoutExporter, err := stdouttrace.New(stdouttrace.WithWriter(f))
+	if err != nil {
+		return nil, err
+	}
+	return &fileExporter{
+		writerExporter: stdoutExporter,
+		file:           f,
+	}, nil
+}
+
+func (e *fileExporter) ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error {
+	return e.writerExporter.ExportSpans(ctx, spans)
+}
+
+func (e *fileExporter) Shutdown(ctx context.Context) error {
+	if err := e.writerExporter.Shutdown(ctx); err != nil {
+		return err
+	}
+	if err := e.file.Close(); err != nil {
+		return fmt.Errorf("closing trace file: %w", err)
+	}
+	return nil
+}
+
+// noopShutdownTracerProvider wraps a TracerProvider with a no-op Shutdown method.
+type noopShutdownTracerProvider struct {
+	tp traceapi.TracerProvider
+}
+
+func (n *noopShutdownTracerProvider) Shutdown(ctx context.Context) error {
+	return nil
+}
+func (n *noopShutdownTracerProvider) Tracer(instrumentationName string, opts ...traceapi.TracerOption) traceapi.Tracer {
+	return n.tp.Tracer(instrumentationName, opts...)
+}
+
+type ShutdownTracerProvider interface {
+	traceapi.TracerProvider
+	Shutdown(ctx context.Context) error
+}
+
+// NewTracerProvider creates and configures a TracerProvider.
+func NewTracerProvider(ctx context.Context) (ShutdownTracerProvider, error) {
+	if os.Getenv("IPFS_TRACING") == "" {
+		return &noopShutdownTracerProvider{tp: traceapi.NewNoopTracerProvider()}, nil
+	}
+
+	options := []trace.TracerProviderOption{}
+
+	traceRatio := 1.0
+	if envRatio := os.Getenv("IPFS_TRACING_RATIO"); envRatio != "" {
+		r, err := strconv.ParseFloat(envRatio, 64)
+		if err == nil {
+			traceRatio = r
+		}
+	}
+	options = append(options, trace.WithSampler(trace.ParentBased(trace.TraceIDRatioBased(traceRatio))))
+
+	r, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String("go-ipfs"),
+			semconv.ServiceVersionKey.String(version.CurrentVersionNumber),
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+	options = append(options, trace.WithResource(r))
+
+	for envVar, builder := range exporterBuilders {
+		if val := os.Getenv(envVar); val != "" {
+			exporter, err := builder(ctx, val)
+			if err != nil {
+				return nil, err
+			}
+			options = append(options, trace.WithBatcher(exporter))
+		}
+	}
+
+	return trace.NewTracerProvider(options...), nil
+}
+
+// Span starts a new span using the standard IPFS tracing conventions.
+func Span(ctx context.Context, componentName string, spanName string, opts ...traceapi.SpanStartOption) (context.Context, traceapi.Span) {
```

Comment:
Yeah that makes sense, I was waiting to create something like this until I used OpenTelemetry in a few repos to understand our usage patterns better. I'm also hesitant to make a tracing "kitchen sink", since it seems like a place that would accumulate dependencies over time and significantly contribute to our dependency hell.